Ical.Net/TimeZone/CalDateTimeZone.cs

```diff
@@ -0,0 +1,101 @@
+//
+// Copyright ical.net project maintainers and contributors.
+// Licensed under the MIT license.
+//
+
+#nullable enable
+using System.Collections.Generic;
+using System.Linq;
+using Ical.Net.CalendarComponents;
+using NodaTime;
+using NodaTime.TimeZones;
+
+namespace Ical.Net.TimeZone;
+
+/// <summary>
+/// Represents a custom <see cref="DateTimeZone"/> that can be used to represent a timezone
+/// created from a <see cref="VTimeZone"/> or to create any other custom time zone.
+/// </summary>
+internal class CalDateTimeZone : DateTimeZone   // for the public API we use DateTimeZone
+{
+    private readonly List<ZoneInterval> _zoneIntervals = new();
+
+    /// <summary>
+    /// Creates a new instance of <see cref="CalDateTimeZone"/> with the specified ID and zone intervals.
+    /// </summary>
+    /// <param name="id">The timezone ID</param>
+    /// <param name="zoneIntervals">A list of <see cref="ZoneInterval"/>s, sorted by start date.</param>
+    public CalDateTimeZone(string id, ICollection<ZoneInterval> zoneIntervals)
+        : base(id, zoneIntervals.Count == 0, GetMinOffset(zoneIntervals), GetMaxOffset(zoneIntervals))
+    {
+        _zoneIntervals.AddRange(zoneIntervals);
+    }
+
+    /// <inheritdoc/>
+    public override ZoneLocalMapping MapLocal(LocalDateTime localDateTime)
+    {
+        // MapLocal calls GetZoneInterval with the localDateTime converted to an Instant
+        var result = base.MapLocal(localDateTime);
+        return result;
+    }
+
+    /// <inheritdoc/>
+    public override ZoneInterval GetZoneInterval(Instant instant)
+    {
+        // instant is a NodaTime internal LocalInstant
+        // that is created from LocalDateTime
+
+        // Handle instants before the first zone interval
+        if (_zoneIntervals.Count == 0 || instant < _zoneIntervals[0].Start)
+            return CreateZoneInterval(
```

Comment:
If we don't have any time zone information, we shouldn't guess here. It's completely unclear, what the reason is for no time zone info being present. It could easily be that the time zone info has intentionally been limited to a certain time range, in which case returning some fallback based on heuristics will most likely be wrong. Just return either null or throw.